### DISCUSSION TITLE
Custom validators based on assigned tags

### DISCUSSION THREAD

**Discussion by Mailstorm-ctrl** [netbox-community/netbox#16198] [Q&A] [answered]
Objective:

If you want to apply a tag with name "Monitored" to a device. this should only complete if:
1. The device has a name <- I know I can get this working so it's not in my conditional
2. The device has a primary IP

I understand the basics of custom validators. I'm halfway there but it seems like I can't get it working. I have:
```
from extras.validators import CustomValidator

class checkMK_TagValidator(CustomValidator):
        def validate(self, instance):
                file1 = open("/opt/netbox/validators/testing/outfile","w")
                file1.write(f"VARS: {vars(instance.tags)}\n")
                file1.write(f"TAGS: {instance.tags}\n")
                file1.write(f"TAG NAMES: {instance.tags.names()}\n")
                #file1.write(f"PIP: {instance.primary_ip}")
                file1.close()
                if "Monitored" in instance.tags.names() and not instance.primary_ip:
                        self.fail("Tag requires a primary IP on the device!", field='tags')
```

When I look at my outfile, I only see the tag name visible after I edit the object a second time (That is, when adding this tag I don't see its name. On a second edit (such as removing the tag), I see the tag name). However, I need this to operate like a normal validator would.

Not entirely sure if this is possible but would be super rad if it was.

**Comment by Mailstorm-ctrl** [comment, marked as answer]
And I figured it out through trial and error:

```
def validate(self, instance):
                if "CheckMK Monitored".upper() in [tag.name.upper() for tag in instance._m2m_values['tags']] and instance.primary_ip is None:
                        self.fail("CheckMK tag requires a primary IP on the device!", field='tags')
```

Not sure if it's the 100% correct way, but it 100% works

**Comment by Azmodeszer** [comment]
I used your solution as a template for another tag-based validation rule of my own (since I could not get [my desired event rule](https://github.com/netbox-community/netbox/discussions/15302) to work):

```
    def validate(self, instance, request):
        if ("webmin-autoupdates".upper() in [tag.slug.upper() for tag in instance._m2m_values['tags']] or "webmin-autosecupdates".upper() in [tag.slug.upper() for tag in instance._m2m_values['tags']]) and not "webmin".upper() in [tag.slug.upper() for tag in instance._m2m_values['tags']]:
            self.fail("Can't add specific Webmin tag without basic Webmin tag!", field='tags')
```

And this works great when I'm editing a single object, but breaks _bulk_ edits:

```
<class 'KeyError'>

'tags'

Python version: 3.11.2
NetBox version: 4.0.9
```
